File: tests/test_answer_refinement.py
"""
Tests for Core/answer_refinement.py — AnswerRefinementEngine

Tests cover:
  - RefinementResult dataclass
  - Score-only mode (no LLM function)
  - refine() async API
  - Completeness check (short answers)
  - to_dict() serialisation
  - Edge cases
"""

import pytest
from unittest.mock import MagicMock


# ─── Fixtures ──────────────────────────────────────────────────────────────


@pytest.fixture
def mock_detector():
    """Fake HallucinationDetector returning clean report."""
    from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport

    det = MagicMock()
    det.detect_hallucinations.return_value = HallucinationReport(
        claims=[ClaimCheck("Claim", None, 0.95)],
        probability=0.05,
        by_type={},
        summary="Clean",
    )
    return det


@pytest.fixture
def engine_score_only():
    """Engine in score-only mode (no LLM, no detector)."""
    from Core.answer_refinement import AnswerRefinementEngine

    return AnswerRefinementEngine(llm_generate_fn=None, hallucination_detector=None)


@pytest.fixture
def engine_with_detector(mock_detector):
    """Engine with mocked detector but no LLM."""
    from Core.answer_refinement import AnswerRefinementEngine

    return AnswerRefinementEngine(
        llm_generate_fn=None,
        hallucination_detector=mock_detector,
        min_answer_words=10,
    )


@pytest.fixture
def engine_full(mock_detector):
    """Engine with mocked LLM + detector."""
    from Core.answer_refinement import AnswerRefinementEngine

    async def fake_llm(prompt):
        return "A revised, longer and more comprehensive answer about the topic."

    return AnswerRefinementEngine(
        llm_generate_fn=fake_llm,
        hallucination_detector=mock_detector,
        min_answer_words=10,
        max_refinement_attempts=1,
    )


# ─── Import ────────────────────────────────────────────────────────────────


def test_import():
    from Core.answer_refinement import AnswerRefinementEngine, RefinementResult

    assert AnswerRefinementEngine is not None
    assert RefinementResult is not None


# ─── RefinementResult ──────────────────────────────────────────────────────


def test_refinement_result_dataclass():
    from Core.answer_refinement import RefinementResult

    r = RefinementResult(
        answer="Revised answer",
        original_answer="Original",
        was_revised=True,
        revision_count=1,
        revision_reasons=["hallucination"],
        quality_score=0.80,
        hallucination_probability=0.15,
        word_count=2,
    )
    assert r.was_revised is True
    assert r.revision_reason == "hallucination"
    assert r.quality_score == 0.80


def test_refinement_result_no_revision():
    from Core.answer_refinement import RefinementResult

    r = RefinementResult(
        answer="Same answer",
        original_answer="Same answer",
        was_revised=False,
        revision_count=0,
        revision_reasons=[],
        quality_score=0.92,
        hallucination_probability=0.02,
        word_count=2,
    )
    assert not r.was_revised
    assert r.revision_reason is None


def test_to_dict():
    from Core.answer_refinement import RefinementResult

    r = RefinementResult(
        answer="Final",
        original_answer="Draft",
        was_revised=True,
        revision_count=1,
        revision_reasons=["too_short"],
        quality_score=0.7,
        hallucination_probability=0.1,
        word_count=1,
    )
    d = r.to_dict()
    assert d["was_revised"] is True
    assert d["quality_score"] == 0.7
    assert "answer" in d
    assert "hallucination_probability" in d


# ─── Score-only refine() ──────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_refine_score_only_mode(engine_score_only):
    """Without LLM/detector, refine should return answer unchanged with quality score."""
    result = await engine_score_only.refine(
        query="What is Python?",
        answer="Python is a versatile programming language used for web development, data science, and machine learning.",
        context="Python is a high-level programming language created by Guido van Rossum.",
    )
    from Core.answer_refinement import RefinementResult

    assert isinstance(result, RefinementResult)
    assert result.answer  # Non-empty
    assert 0.0 <= result.quality_score <= 1.0


@pytest.mark.asyncio
async def test_refine_preserves_good_answer(engine_with_detector):
    """Good answer that passes checks should remain unchanged."""
    good_answer = (
        "Python was created by Guido van Rossum and first released in 1991. "
        "It supports multiple paradigms including OOP, functional, and procedural."
    )
    result = await engine_with_detector.refine(
        query="What is Python?",
        answer=good_answer,
        context="Python was created by Guido van Rossum. First released in 1991.",
    )
    assert result.answer == good_answer
    assert not result.was_revised


# ─── With detector ────────────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_refine_calls_detector(engine_with_detector, mock_detector):
    """Detector should be called during refinement."""
    await engine_with_detector.refine(
        query="test",
        answer="A detailed answer with more than ten words for the completeness check.",
        context="context",
    )
    mock_detector.detect_hallucinations.assert_called()


# ─── Short answer completeness ────────────────────────────────────────────


@pytest.mark.asyncio
async def test_short_answer_detected(engine_full):
    """Very short answer should be detected as incomplete."""
    result = await engine_full.refine(
        query="Explain quantum computing in detail",
        answer="It uses qubits.",
        context="Quantum computing uses qubits which can be in superposition states.",
    )
    # The engine should try to expand it via the LLM
    from Core.answer_refinement import RefinementResult

    assert isinstance(result, RefinementResult)


# ─── Edge Cases ────────────────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_empty_answer(engine_score_only):
    """Empty answer should not crash."""
    result = await engine_score_only.refine("query", "", "context")
    assert isinstance(result.quality_score, float)


@pytest.mark.asyncio
async def test_empty_context(engine_score_only):
    """Empty context should not crash."""
    result = await engine_score_only.refine("query", "Some answer text here with enough words.", "")
    assert isinstance(result.quality_score, float)

File: rust_output/src/test_answer_refinement.rs
/// Tests for Core/answer_refinement::py — AnswerRefinementEngine
/// 
/// Tests cover:
/// - RefinementResult dataclass
/// - Score-only mode (no LLM function)
/// - refine() async API
/// - Completeness check (short answers)
/// - to_dict() serialisation
/// - Edge cases

use anyhow::{Result, Context};
use std::collections::HashMap;
use tokio;

/// Fake HallucinationDetector returning clean report.
pub fn mock_detector() -> () {
    // Fake HallucinationDetector returning clean report.
    // TODO: from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport
    let mut det = MagicMock();
    det.detect_hallucinations.return_value = HallucinationReport(/* claims= */ vec![ClaimCheck("Claim".to_string(), None, 0.95_f64)], /* probability= */ 0.05_f64, /* by_type= */ HashMap::new(), /* summary= */ "Clean".to_string());
    det
}

/// Engine in score-only mode (no LLM, no detector).
pub fn engine_score_only() -> () {
    // Engine in score-only mode (no LLM, no detector).
    // TODO: from Core.answer_refinement import AnswerRefinementEngine
    AnswerRefinementEngine(/* llm_generate_fn= */ None, /* hallucination_detector= */ None)
}

/// Engine with mocked detector but no LLM.
pub fn engine_with_detector(mock_detector: String) -> () {
    // Engine with mocked detector but no LLM.
    // TODO: from Core.answer_refinement import AnswerRefinementEngine
    AnswerRefinementEngine(/* llm_generate_fn= */ None, /* hallucination_detector= */ mock_detector, /* min_answer_words= */ 10)
}

/// Engine with mocked LLM + detector.
pub fn engine_full(mock_detector: String) -> () {
    // Engine with mocked LLM + detector.
    // TODO: from Core.answer_refinement import AnswerRefinementEngine
    let fake_llm = |prompt| {
        "A revised, longer and more comprehensive answer about the topic.".to_string()
    };
    AnswerRefinementEngine(/* llm_generate_fn= */ fake_llm, /* hallucination_detector= */ mock_detector, /* min_answer_words= */ 10, /* max_refinement_attempts= */ 1)
}

pub fn test_import() -> () {
    // TODO: from Core.answer_refinement import AnswerRefinementEngine, RefinementResult
    assert!(AnswerRefinementEngine.is_some());
    assert!(RefinementResult.is_some());
}

pub fn test_refinement_result_dataclass() -> () {
    // TODO: from Core.answer_refinement import RefinementResult
    let mut r = RefinementResult(/* answer= */ "Revised answer".to_string(), /* original_answer= */ "Original".to_string(), /* was_revised= */ true, /* revision_count= */ 1, /* revision_reasons= */ vec!["hallucination".to_string()], /* quality_score= */ 0.8_f64, /* hallucination_probability= */ 0.15_f64, /* word_count= */ 2);
    assert!(r.was_revised == true);
    assert!(r.revision_reason == "hallucination".to_string());
    assert!(r.quality_score == 0.8_f64);
}

pub fn test_refinement_result_no_revision() -> () {
    // TODO: from Core.answer_refinement import RefinementResult
    let mut r = RefinementResult(/* answer= */ "Same answer".to_string(), /* original_answer= */ "Same answer".to_string(), /* was_revised= */ false, /* revision_count= */ 0, /* revision_reasons= */ vec![], /* quality_score= */ 0.92_f64, /* hallucination_probability= */ 0.02_f64, /* word_count= */ 2);
    assert!(!r.was_revised);
    assert!(r.revision_reason.is_none());
}

pub fn test_to_dict() -> () {
    // TODO: from Core.answer_refinement import RefinementResult
    let mut r = RefinementResult(/* answer= */ "Final".to_string(), /* original_answer= */ "Draft".to_string(), /* was_revised= */ true, /* revision_count= */ 1, /* revision_reasons= */ vec!["too_short".to_string()], /* quality_score= */ 0.7_f64, /* hallucination_probability= */ 0.1_f64, /* word_count= */ 1);
    let mut d = r.to_dict();
    assert!(d["was_revised".to_string()] == true);
    assert!(d["quality_score".to_string()] == 0.7_f64);
    assert!(d.contains(&"answer".to_string()));
    assert!(d.contains(&"hallucination_probability".to_string()));
}

/// Without LLM/detector, refine should return answer unchanged with quality score.
pub async fn test_refine_score_only_mode(engine_score_only: String) -> () {
    // Without LLM/detector, refine should return answer unchanged with quality score.
    let mut result = engine_score_only.refine(/* query= */ "What is Python?".to_string(), /* answer= */ "Python is a versatile programming language used for web development, data science, and machine learning.".to_string(), /* context= */ "Python is a high-level programming language created by Guido van Rossum.".to_string()).await;
    // TODO: from Core.answer_refinement import RefinementResult
    assert!(/* /* isinstance(result, RefinementResult) */ */ true);
    assert!(result.answer);
    assert!((0.0_f64 <= result.quality_score) && (result.quality_score <= 1.0_f64));
}

/// Good answer that passes checks should remain unchanged.
pub async fn test_refine_preserves_good_answer(engine_with_detector: String) -> () {
    // Good answer that passes checks should remain unchanged.
    let mut good_answer = "Python was created by Guido van Rossum and first released in 1991. It supports multiple paradigms including OOP, functional, and procedural.".to_string();
    let mut result = engine_with_detector.refine(/* query= */ "What is Python?".to_string(), /* answer= */ good_answer, /* context= */ "Python was created by Guido van Rossum. First released in 1991.".to_string()).await;
    assert!(result.answer == good_answer);
    assert!(!result.was_revised);
}

/// Detector should be called during refinement.
pub async fn test_refine_calls_detector(engine_with_detector: String, mock_detector: String) -> () {
    // Detector should be called during refinement.
    engine_with_detector.refine(/* query= */ "test".to_string(), /* answer= */ "A detailed answer with more than ten words for the completeness check.".to_string(), /* context= */ "context".to_string()).await;
    mock_detector.detect_hallucinations.assert_called();
}

/// Very short answer should be detected as incomplete.
pub async fn test_short_answer_detected(engine_full: String) -> () {
    // Very short answer should be detected as incomplete.
    let mut result = engine_full.refine(/* query= */ "Explain quantum computing in detail".to_string(), /* answer= */ "It uses qubits.".to_string(), /* context= */ "Quantum computing uses qubits which can be in superposition states.".to_string()).await;
    // TODO: from Core.answer_refinement import RefinementResult
    assert!(/* /* isinstance(result, RefinementResult) */ */ true);
}

/// Empty answer should not crash.
pub async fn test_empty_answer(engine_score_only: String) -> () {
    // Empty answer should not crash.
    let mut result = engine_score_only.refine("query".to_string(), "".to_string(), "context".to_string()).await;
    assert!(/* /* isinstance(result.quality_score, float) */ */ true);
}

/// Empty context should not crash.
pub async fn test_empty_context(engine_score_only: String) -> () {
    // Empty context should not crash.
    let mut result = engine_score_only.refine("query".to_string(), "Some answer text here with enough words.".to_string(), "".to_string()).await;
    assert!(/* /* isinstance(result.quality_score, float) */ */ true);
}
